hardware_interface/src/resource_manager.cpp

```diff
@@ -0,0 +1,248 @@
+// Copyright 2020 ROS2-Control Development Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "hardware_interface/resource_manager.hpp"
+
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "hardware_interface/actuator_hardware_interface.hpp"
+#include "hardware_interface/actuator_hardware.hpp"
+#include "hardware_interface/component_parser.hpp"
+#include "hardware_interface/hardware_info.hpp"
+#include "hardware_interface/robot_hardware_interface.hpp"
+#include "hardware_interface/sensor_hardware_interface.hpp"
+#include "hardware_interface/sensor_hardware.hpp"
+#include "hardware_interface/system_hardware_interface.hpp"
+#include "hardware_interface/system_hardware.hpp"
+#include "hardware_interface/types/hardware_interface_return_values.hpp"
+#include "pluginlib/class_loader.hpp"
+#include "rclcpp/rclcpp.hpp"
+
+namespace
+{
+constexpr const auto kLoggerName = "ros2_control_ResourceManager";
+}  // namespace
+
+namespace resource_manager
+{
+
+ResourceManager::ResourceManager()
+{
+  actuator_loader_.reset(
+    new pluginlib::ClassLoader<hardware_interface::ActuatorHardwareInterface>(
+      "hardware_interface", "hardware_interface::ActuatorHardwareInterface"));
+  sensor_loader_.reset(
+    new pluginlib::ClassLoader<hardware_interface::SensorHardwareInterface>(
+      "hardware_interface", "hardware_interface::SensorHardwareInterface"));
+  system_loader_.reset(
+    new pluginlib::ClassLoader<hardware_interface::SystemHardwareInterface>(
+      "hardware_interface", "hardware_interface::SystemHardwareInterface"));
+
+  joint_loader_.reset(
+    new pluginlib::ClassLoader<hardware_interface::components::Joint>(
+      "hardware_interface", "hardware_interface::components::Joint"));
```

Comment:
consider to put them in the initializer list with `std::make_shared` / `std::make_unique`.
But first, why are these shared pointer and not unique pointers, and second, why are these pointers in the first place?